worksheet.rb

```diff
@@ -23,12 +56,98 @@
 # into this data structure, such as "DR0004"
 # and "3rd Feb 2016" and "RD0022"
 
-########################################################
+#######################################################
 # Step 4: Total Driver's Earnings and Number of Rides
-
 # Use an iteration blocks to print the following answers:
+
+def line
+  puts "#######################################################"
+end
+
+line
+
 # - the number of rides each driver has given
+
+def num_ride(ride_share)
+  ride_share.each do |key, value|
+    puts "#{key} has #{value.length} rides."
+    end
+end
+
+num_ride(ride_share)
+
+line
+
 # - the total amount of money each driver has made
+
+def driver_earn(ride_share)
+  ride_share.each do | driver, array |
+    total = 0
+     array.each do |ride|
+       total += ride[:cost]
+    end
+    puts "#{driver} made $#{total}"
+  end
+end
+
+driver_earn(ride_share)
+
+line
+
 # - the average rating for each driver
+
+def driver_avg(ride_share)
+  ride_share.each do | driver, array |
+    total = 0
+     array.each do |ride|
+       total += ride[:rating]
+     end
+    average = total.to_f / array.length
+    puts "#{driver} made #{average.round(1)}"
+  end
+end
+
+driver_avg(ride_share)
+
+line
+
 # - Which driver made the most money?
-# - Which driver has the highest average rating?
\ No newline at end of file
+
+def driver_max_earn(ride_share)
+  totals_hash = {}
+  ride_share.each do | driver, array |
+    total = 0
+     array.each do |ride|
+       total += ride[:cost]
+    end
+    totals_hash[driver] = total
+    # max_arr << totals_hash
+  end
+  return totals_hash
+end
+
+highest_earner = driver_max_earn(ride_share).max_by { | driver, ride_totals | ride_totals }
+puts "#{highest_earner[0]} earned the most with $#{highest_earner[1]}, BALLIN'!!!"
+
+line
+
+# - Which driver has the highest average rating?eb 2016", cost: "5", rating: "5"}]}
+
+def driver_max_rate(ride_share)
+  totals_rating_hash = {}
+  ride_share.each do | driver, array |
+    total = 0
+    array.each do |ride|
+      total += ride[:rating]
+    end
+    average = total.to_f / array.length
+    totals_rating_hash[driver] = average.round(1)
+  end
+  return totals_rating_hash
+end
+
+highest_rating = driver_max_rate(ride_share).max_by { | driver, rate_totals| rate_totals}
+
+puts "#{highest_rating[0]} had the highest rating with #{highest_rating[1]}"
```

Comment:
Consider organizing your program so that all method definitions are grouped together, and all code that is focused on outputs for the user are grouped to increase readability.